Jenkins test failure for test StreamSegmentContainerTests#testTransactionOperations

**Problem description**
Jenkins test failure : `StreamSegmentContainerTests#testTransactionOperations`

```Error Msg:
java.util.concurrent.ExecutionException: io.pravega.common.ObjectClosedException: Object 'FutureReadResultEntryCollection' has been closed and cannot be accessed anymore.

Stacktrace

java.util.concurrent.ExecutionException: io.pravega.common.ObjectClosedException: Object 'FutureReadResultEntryCollection' has been closed and cannot be accessed anymore.
    at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
    at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
    at io.pravega.segmentstore.server.containers.StreamSegmentContainerTests.mergeTransactions(StreamSegmentContainerTests.java:1237)
    at io.pravega.segmentstore.server.containers.StreamSegmentContainerTests.testTransactionOperations(StreamSegmentContainerTests.java:638)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.lang.Thread.run(Thread.java:745)
Caused by: io.pravega.common.ObjectClosedException: Object 'FutureReadResultEntryCollection' has been closed and cannot be accessed anymore.
    at io.pravega.common.Exceptions.checkNotClosed(Exceptions.java:163)```

**Problem location**

StreamSegmentContainerTests#testTransactionOperations

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
@@ -206,13 +206,15 @@ public class ContainerReadIndex implements ReadIndex {
                 if (getIndex(segmentId) == null) {
                     throw ex;
                 } else {
-                    log.debug("{}: triggerFutureReads: StreamSegmentId {} was skipped because it is no longer registered.", this.traceObjectId, segmentId);
+                    log.debug("{}: triggerFutureReads: StreamSegmentId {} was skipped because it is no longer registered.",
+                            this.traceObjectId, segmentId);
                 }
             }
         }
 
         // Throw any exception at the end - we want to make sure at least the ones that did have a valid index entry got triggered.
-        Exceptions.checkArgument(missingIds.size() == 0, "streamSegmentIds", "At least one StreamSegmentId does not exist in the metadata: %s", missingIds);
+        Exceptions.checkArgument(missingIds.size() == 0, "streamSegmentIds",
+                "At least one StreamSegmentId does not exist in the metadata: %s", missingIds);
     }
 
     @Override

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
@@ -11,6 +11,7 @@ package io.pravega.segmentstore.server.reading;
 
 import com.google.common.base.Preconditions;
 import io.pravega.common.Exceptions;
+import io.pravega.common.ObjectClosedException;
 import io.pravega.segmentstore.contracts.ReadResult;
 import io.pravega.segmentstore.server.ContainerMetadata;
 import io.pravega.segmentstore.server.DataCorruptionException;
@@ -189,8 +190,24 @@ public class ContainerReadIndex implements ReadIndex {
                         missingIds.add(segmentId);
                     }
                 }
-            } else {
+                continue;
+            }
+
+            try {
                 index.triggerFutureReads();
+            } catch (ObjectClosedException ex) {
+                // It is possible that between the time we got the pointer to the StreamSegmentReadIndex and when we got
+                // to invoking triggerFutureReads, the StreamSegmentReadIndex has already been closed. If this is the case,
+                // ignore the error.
+                // This is possible in the following scenario: for a Transaction, we have an Append, followed by a Merge;
+                // the Append makes this index eligible for triggering future reads, and the Merge (once committed to Storage)
+                // will close it. If the StorageWriter is sufficiently fast in comparison to the OperationProcessor callbacks
+                // (which could be the case for in-memory unit tests), it may trigger this condition.
+                if (getIndex(segmentId) == null) {
+                    throw ex;
+                } else {
+                    log.debug("{}: triggerFutureReads: StreamSegmentId {} was skipped because it is no longer registered.", this.traceObjectId, segmentId);
+                }
             }
         }
 

